### PR TITLE
[Fix] Setting batch after finalize/initialize

### PR DESCRIPTION
**Pull request by zhoonit**
## Commits to be reviewed in this PR

<details><summary>[Fix] Setting batch after finalize/initialize</summary><br />

This patch delays setting batch after finalize/initialize.

Setting batch must be done after runcontext has made, the reason is that
we have semantics that finalize should be independent of batch size, but
if batch size is set before output_dims batch must be set according to
input_dims batch, which is not desirable.

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

</details>

### CODE REVIEW

File: nntrainer/models/neuralnet.cpp
Comment (kparichay) [inline nntrainer/models/neuralnet.cpp:194]: :100:

**Review by kparichay — APPROVED** [review]
LGTM after CI fix
You might need to fix bugs in some layer if setBatch() is not properly supported by them

**Comment by zhoonit** [comment]
> LGTM after CI fix You might need to fix bugs in some layer if setBatch() is not properly supported by them

Yeah, there were some (hidden) bugs related to this. Please take another look :)

File: nntrainer/layers/layer_node.cpp
Comment (zhoonit) [inline nntrainer/layers/layer_node.cpp:597]: @note: we don't (and should not) query input shape property for querying (current and actual) input dimension, by this we can segregate + detect bugs earlier which would happen by assuming input_shape has the batch dimension as well.

File: nntrainer/layers/rnncell.cpp
Comment (zhoonit) [inline nntrainer/layers/rnncell.cpp:257]: FYI) @lhs8928

File: nntrainer/layers/bn_layer.cpp
Comment (kparichay) [inline nntrainer/layers/bn_layer.cpp:263]: This is already done by layer node and can be skipped.

File: nntrainer/graph/network_graph.cpp
Comment (kparichay) [inline nntrainer/graph/network_graph.cpp:402]: Are you sure to move it here? If this is moved here, then the changes requested here by the layers will not be reflected (which is done from line 392 to 399)

Comment (zhoonit) [reply nntrainer/graph/network_graph.cpp:402]: Yeah, this is intended. 
for example below should be override default behavior.
```cpp
  context.updateTensor(wt_idx[RNNCellParams::hidden_state],
                       batch * max_timestep);
```

Please correct me if I got something wrong :)

Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:402]: You are right that context.updateTensor() should decide how to update the tensor shapes, but IMU this updated information should reach tensor_pool, which is done from L392 to L399.

Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:402]: The memory is always de-allocated with setBatch(), so `context.updateTensor()` alone is sufficient.
L392 to L399 can be removed IMO as its purpose was to pass the information to the tensor_pool if not deallocated, please check.

Comment (zhoonit) [reply nntrainer/graph/network_graph.cpp:402]: Yeah, the function guarantee's it is deallocated so it should be fine. But I'll delay removing the part as we don't have a proper test scenario that batch is being changed and run train or inference. (this will have to be added to a model test scenario). I am suspecting some layers requires setBatch() but did not create one, I'll leave an issue instead for now.

Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:402]: We do have a model test with ccapi :) - https://github.com/nnstreamer/nntrainer/blob/c09bbf73ae234e9c3bd989c59d3fb7af220679b7/test/ccapi/unittest_ccapi.cpp#L364-L386

Comment (zhoonit) [reply nntrainer/graph/network_graph.cpp:402]: Awesome :)

File: nntrainer/layers/bn_layer.cpp
Comment (zhoonit) [reply nntrainer/layers/bn_layer.cpp]: Please note that, input_dim.batch is not updated when this function is being called. We might want to update inputdims before coming here later.

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Pull request merged**
